sdk/detect/auto.go

```diff
@@ -0,0 +1,41 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package detect
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel/sdk/resource"
+)
+
+// ResourceDetector attempts to detect resource information.
+// If the detector cannot find resource information, the returned resource is nil but no
+// error is returned.
+// An error is only returned on unexpected failures.
+type ResourceDetector func(ctx context.Context) (*resource.Resource, error)
```

Comment:
I don't know if it's just because of coming from a more OO background, but personally I would prefer this to be an interface:

```go
type Detector interface {
    func Detect(ctx context.Context) (*resource.Resource, error)
}
```

I'm not actually sure which is more idiomatic in Go, but these kind of interfaces are at least common as per https://golang.org/doc/effective_go.html#interface-names

Reply:
Yup, using an interface here allows multiple separate implementations that can live in separate packages even but still be consumed as a single type.  In Go it is common to declare interfaces on the consuming side as any type that implements the methods declared in the interface will satisfy the interface.  So, here we would define the `Detector` where it is used in the `Detect()` function and users would be able to provide implementations of the `Detector` interface from any package they choose.